pkg/scaling/scale_handler.go

```diff
@@ -576,81 +587,190 @@ func (h *scaleHandler) getScaledObjectState(ctx context.Context, scaledObject *k
 
-			metricName := spec.External.Metric.Name
-
-			var latency int64
-			metrics, isMetricActive, latency, err := cache.GetMetricsAndActivityForScaler(ctx, scalerIndex, metricName)
-			if latency != -1 {
-				prommetrics.RecordScalerLatency(scaledObject.Namespace, scaledObject.Name, scalerName, scalerIndex, metricName, float64(latency))
+			for _, metric := range matchingMetrics {
+				value := metric.Value.AsApproximateFloat64()
+				prommetrics.RecordScalerMetric(scaledObject.Namespace, scaledObject.Name, kedav1alpha1.CompositeMetricName, 0, metric.MetricName, value)
+				prommetrics.RecordScalerActive(scaledObject.Namespace, scaledObject.Name, kedav1alpha1.CompositeMetricName, 0, metric.MetricName, value > activationValue)
+				if !isScaledObjectActive {
+					isScaledObjectActive = value > activationValue
+				}
 			}
-			logger.V(1).Info("Getting metrics and activity from scaler", "scaler", scalerName, "metricName", metricName, "metrics", metrics, "activity", isMetricActive, "scalerError", err)
+		}
+	}
 
-			if scalerConfigs[scalerIndex].TriggerUseCachedMetrics {
-				metricsRecord[metricName] = metricscache.MetricsRecord{
-					IsActive:    isMetricActive,
-					Metric:      metrics,
-					ScalerError: err,
-				}
+	// cpu/memory scaler only can scale to zero if there is any other external metric because otherwise
+	// it'll never scale from 0. If all the triggers are only cpu/memory, we enforce the IsActive
+	if len(scaledObject.Spec.Triggers) <= cpuMemCount && !isScaledObjectError {
+		isScaledObjectActive = true
+	}
+	return isScaledObjectActive, isScaledObjectError, metricsRecord, err
+}
+
+// getStateScalerResult is used as paremeter
+// for the function getScalerState. It contains
+// all the needed information for getting the
+// scaler state
+type getStateScalerRequest struct {
+	ctx          context.Context
+	scaler       scalers.Scaler
+	scalerConfig scalers.ScalerConfig
+	scalerIndex  int
+	cache        *cache.ScalersCache
+	logger       logr.Logger
+	scaledObject *kedav1alpha1.ScaledObject
+}
+
+// getStateScalerResult is used as return
+// for the funtion getScalerState. It contains
+// the state of the scaler and all the required
+// info for calculating the ScaledObjectState
+type getStateScalerResult struct {
+	// IsActive will be overrided by formula calculation
+	IsActive bool
+	IsError  bool
+	Metrics  []external_metrics.ExternalMetricValue
+	Pairs    map[string]string
+	Records  map[string]metricscache.MetricsRecord
+}
+
+// getScalerState returns getStateScalerResult with the state
+// for an specific request (getStateScalerRequest). The state
+// contains if it's active or with erros, but also the records
+// for the cache and he metrics for the custom formulas
+func (*scaleHandler) getScalerState(
+	request getStateScalerRequest,
+) getStateScalerResult {
+	result := getStateScalerResult{
+		IsActive: false,
+		IsError:  false,
+		Metrics:  []external_metrics.ExternalMetricValue{},
+		Pairs:    map[string]string{},
```

Comment:
Not relevant to this PR.

But for the sake of simplicity, can't we somehow get store the same trigger-name information in the already existing Records struct? @gauron99

Reply:
I plan to revisit this with the parallel PR, but for not locking the release I just added all the used members

Reply:
Yeah, absolutely. I was talking about the original design.

Reply:
im not sure about the Records struct TBH I will take a look into that